MicroProfile Metrics 5.0 documentation updates. (#6180)
OpenLiberty/docs: modules/ROOT/pages/microservice-observability-metrics.adoc

You can write an application to produce metrics that operations teams use when the application is running in production.
When an application is running, you can view your metric data from any browser by accessing a `/metrics` endpoint, for example, `https://localhost:9443/metrics`, where `9443` is the port number for your application.

> You can narrow down the scope of the metrics by using the `scope` query parameter with the `/metrics` endpoint to specify the metric registry scope. For example, to view the metric data for the `base` metric registry, use the `/metrics?scope=base` endpoint. This pattern is used to access all runtime-provided scopes,  (`base`,`application`, and `vendor`) and any <<customscope, user-defined metric registry scopes>>. +

Reviewer: ```suggestion
You can narrow the scope of the metrics by using the `scope` query parameter with the `/metrics` endpoint to specify the metric registry scope. For example, to view the metric data for the `base` metric registry, use the `/metrics?scope=base` endpoint. This pattern is used to access all runtime-provided scopes  (`base`,`application`, and `vendor`) and any <<customscope, user-defined metric registry scopes>>. +
```

> You can narrow down the scope of the metrics by using the "scope" query parameter with the `/metrics` endpoint to specify the metric registry scope. For example, to view the metric data for the `base` metric registry, use the `/metrics?scope=base` endpoint. This pattern is used to access all runtime provided scopes (i.e. `base`,`application` and `vendor`) and any <<customscope,  user-defined metric registry scopes>>. +

Reviewer: ```suggestion
You can narrow down the scope of the metrics by using the `scope` query parameter with the `/metrics` endpoint to specify the metric registry scope. For example, to view the metric data for the `base` metric registry, use the `/metrics?scope=base` endpoint. This pattern is used to access all runtime-provided scopes,  (`base`,`application`, and `vendor`) and any <<customscope, user-defined metric registry scopes>>. +
```

> **For MicroProfile Metrics 4.0 and earlier**, only runtime provided scopes are available and path parameters are used to narrow down metric data. The endpoints are `/metrics?scope=base`, `/metrics?scope=application`, and `/metrics?scope=vendor`.

Reviewer: ```suggestion
For MicroProfile Metrics 4.0 and earlier, only runtime-provided scopes are available and path parameters are used to narrow down metric data. The endpoints are `/metrics?scope=base`, `/metrics?scope=application`, and `/metrics?scope=vendor`.
```

Reviewer: using bold in this context will draw accessibility flags

Metric data is emitted in https://prometheus.io/docs/instrumenting/exposition_formats/[Prometheus format].
Operations teams can gather the metrics and store them in a database by using tools like Prometheus.
They can then visualize the metric data for analysis in dashboards such as https://grafana.com/[Grafana].
For a list of metrics that are available in Open Liberty, see the xref:metrics-list.adoc[metrics reference list].

> **MicroProfile Metrics runtimes that are version 4.0 and earlier** can also export metric data in JSON format through the `/metrics` endpoints. JSON formatted metric data can be obtained by confiuring the `Accept` header of your request to the `application/json` value.

Reviewer: ```suggestion
MicroProfile Metrics runtimes version 4.0 and earlier can also export metric data in JSON format through the `/metrics` endpoints. JSON-formatted metric data can be obtained by configuring the `Accept` header of your request to the `application/json` value.
```

A `GET` request returns a list of metrics, and an `OPTIONS` request returns a list of metrics with their metadata.

== Adding metrics to your applications

You can use MicroProfile Metrics with Open Liberty by enabling the feature:mpMetrics[display=MicroProfile Metrics] feature in your `server.xml` file.
To add metrics to your applications, you must create and register metrics with the default application registry or your own <<customscope,  user-defined metric registry scope>> so that they're known to the system and can be reported on from the `/metrics` endpoint.
The simplest way to add metrics to your application is by using annotations.
link:/docs/ref/microprofile/3.3/#package=org/eclipse/microprofile/metrics/annotation/package-frame.html&class=org/eclipse/microprofile/metrics/annotation/package-summary.html[MicroProfile Metrics defines annotations] that you can use to quickly build metrics into your code.
These metrics ultimately provide transparency for operations teams into how services are running.

[#customscope]
== User-defined metric registry scopes

> Starting in MicroProfile Metrics 5.0, in addition to the default `application` metric registry scope, you are now able to define your own metric registry scope to register metrics to. This can be achieved by specifying the `scope` parameter for metric annotations or by injecting a `MetricRegistry` with the `RegistryScope` annotation and specifying the `scope` parameter. Not defining a value for the `scope` parameter in the metric annotation or `RegistryScope` annotation will default to the `application` scope.

Reviewer: ```suggestion
Starting in MicroProfile Metrics 5.0, in addition to the default `application` metric registry scope, you can define your own metric registry scope to register metrics to. You can define your own scope by specifying the `scope` parameter for metric annotations or by injecting a `MetricRegistry` with the `RegistryScope` annotation and specifying the `scope` parameter. If you don't define a value for the `scope` parameter in the metric annotation or `RegistryScope` annotation, the default scope is `application`.
```

> Example using the `@Counted` metric annotation to define a user-defined scope

Reviewer: ```suggestion
The following example uses the `@Counted` metric annotation to define a user-defined scope.
```

----
@Counted(name="customCounter" description="This counter will preside in a user-defined metric registry", scope="customScope")
public String countingThings() {
    return "1 2 3 4..!";
}
----

> Example injecting a `MetricRegistry` for instrumenting metrics with the Java API.

Reviewer: ```suggestion
The following example injects a `MetricRegistry` for instrumenting metrics with the Java API.
```

----
@Inject
@RegistryScope(scope="customScope")
MetricRegistry customMetricRegistry;
----

== Metrics and annotations
The following sections describe metric types that are available and how their corresponding annotations are used.

> The following list lists metric types available in the **latest** MicroProfile Metrics feature.

Reviewer: ```suggestion
The following metric types are available when the MicroProfile Metrics  5.0 feature is enabled.
```

> The following metric types are available when the MicroProfile Metrics  5.0 feature is enabled.

Reviewer: ```suggestion
The following metric types are available when the MicroProfile Metrics 5.0 feature is enabled.
```

* <<counter,Counter>>
* <<gauge,Gauge>>
* <<meter,Meter>>
* <<timer,Timer>>
* <<histogram,Histogram>>

> Throughout the many releases of MicroProfile Metrics different metric types were introduced, removed or have had their behavior changed.

Reviewer: ```suggestion
Throughout the many releases of MicroProfile Metrics different metric types were introduced, removed, or modified.
```

> The following list lists metric types available in previous MicroProfile Metrics releases. Metric types that are found both lists indicates that its behavior and/or capability has changed.

Reviewer: ```suggestion
The following list describes metric types that are available in previous MicroProfile Metrics releases. Metrics that are listed in both MicroProfile Metrics 5.0 and earlier releases behave differently before version 5.0.
```

}
----

[#timer]
=== Timer
A timer metric aggregates timing durations in nanoseconds and provides duration statistics.

Use the `@Timed` annotation to mark a constructor or method as a timer.

> The timer tracks how frequently the annotated object is started and how long the invocations take to complete:

Reviewer: ```suggestion
The timer tracks how frequently the annotated object is started and how long the invocations take to complete, as shown in the following example.
```

[source,java]
----
@POST
@Path("/creditcard")
@Timed(
    name="donateAmountViaCreditCard.timer",
    description = "Donations that were made using a credit card")
public String donateAmountViaCreditCard(@FormParam("amount") Long amount, @FormParam("card") String card) {

    if (processCard(card, amount))
        return "Thanks for donating!";

    return "Sorry, please try again.";
}
----

[#histogram]
=== Histogram
A histogram is a metric that calculates the distribution of a value. It provides the following information:

- Maximum, median and mean values
- The value at the 50th, 75th, 95th, 98th, 99th, 99.9th percentile
- A count of the number of values

> Note: When viewing the Prometheus formatted metric data for a histogram, the mean value is not included.

Reviewer: ```suggestion
Note: When you view the Prometheus-formatted metric data for a histogram, the mean value is not included.
```

> The histogram metric does not have an annotation. To record a value in the histogram, you must call the `histogram.update(long value)` method with the value that you want to record. The current state, or snapshot, of the histogram can be retrieved by using the `getSnapshot()` method. Histograms in MicroProfile Metrics only support integer or long values.

Reviewer: ```suggestion
The histogram metric does not have an annotation. To record a value in the histogram, you must call the `histogram.update(long value)` method with the value that you want to record. The current state, or snapshot, of the histogram can be retrieved by using the `getSnapshot()` method. Histograms in MicroProfile Metrics support only integer or long values.
```

> Examples of histograms include the distribution of payload sizes that are retrieved or for an onboarding survey that collects the distribution of household income.

Reviewer: ```suggestion
An example of a histogram is the distribution of payload sizes that are retrieved for an onboarding survey that collects the distribution of household income.
```

donationDistribution_Dollars_count{mp_scope="application",tier="integration",} 203.0
donationDistribution_Dollars_sum{mp_scope="application",tier="integration",} 91850.0
# HELP donationDistribution_Dollars_max The distribution of the donation amounts
# TYPE donationDistribution_Dollars_max gauge
donationDistribution_Dollars_max{mp_scope="application",tier="integration",} 1000.0
----

'''

[#concurrentgaugeOld]
=== Concurrent gauge [MicroProfile Metrics 2.0-4.0]
A concurrent gauge metric counts the concurrent invocations of an annotated element.
This metric also tracks the high and low watermarks of each invocation.

Use the `@ConcurrentGauge` annotation to mark a method as a concurrent gauge.
The concurrent gauge increments when the annotated method is called and decrements when the annotated method returns, counting current invocations of the annotated method:

[source,java]
----
@GET
@Path("/livestream");
@ConcurrentGauge(name = "liveStreamViewers", displayName="Donation live stream viewers", description="Number of active viewers for the donation live stream")
public void donationLiveStream() {
    launchLiveStreamConnection();
}
----

[#meter]

> === Meter [MicroProfile Metrics 1.0-4.0]

Reviewer: ```suggestion
=== Meter (available only in MicroProfile Metrics 1.0-4.0)
```

----

[#simpletimer]

> === Simple timer [MicroProfile Metrics 3.0-4.0]

Reviewer: ```suggestion
=== Simple timer (available only in MicroProfile Metrics 3.0-4.0)
```

}
----

[#timerOld]

> === Timer [MicroProfile Metrics 1.0-4.0]

Reviewer: ```suggestion
=== Timer (available only in MicroProfile Metrics 1.0-4.0)
```

}
----

[#histogramOld]

> === Histogram [MicroProfile Metrics 1.0-4.0]

Reviewer: ```suggestion
=== Histogram (available only in MicroProfile Metrics 1.0-4.0)
```

'''

These types of metrics are available to add to your applications to make them observable.
In production, operations teams can use these metrics to monitor the application, along with metrics that are automatically emitted from the JVM and the Open Liberty server runtime.
If you're interested in learning more about using MicroProfile Metrics to build observability into your microservices, see the Open Liberty guide for link:/guides/microprofile-metrics.html[Providing metrics from a microservice].

'''

== Metric Tags

> Metrics instrumented with MicroProfile Metrics can be instrumented with key-value tags. The key is the tag name and the value can be referred to as the tag value. This allows for the instrumentation of multiple metrics, of the same type, using the same metric name as long as the set of tags have matching tag names, but with different values. If the metric types used or if the set of tag names do not match, then an `IllegalArgumentException` will be thrown. This is useful for logically grouping metrics that track similar statistics by using the same name and are then identified by their tags.

Reviewer: this paragraph is really confusing, though i get the basic idea

Author: How about if it was trimmed down to:

Metrics instrumented with MicroProfile Metrics can be instrumented with tags. The tag consists of a name and value. This allows for the instrumentation of multiple metrics, of the same type, using the same metric name as long as the set of tags have matching names, but with different values. If the metric types used or if the set of tag names do not match, then an `IllegalArgumentException` will be thrown. This is useful for logically grouping metrics that track similar statistics by using the same name and are then identified by their tags.

Reviewer: what if we lead with the benefit- something like:

> With MicroProfile metrics, you can logically group metrics that track similar statistics by instrumenting them with tags. Tags consist of a name and a value. You can instrument multiple metrics of the same type with tags of the same name but different values. However, in this configuration, if the metrics types are not the same or if the tag names do not match exactly, an `IllegalArgumentException` is thrown.

> For example, using counters, we can create multiple counters named "methodCounter" which returns the count of the method identified by the tag.

Reviewer: ```suggestion
For example, when you use counters, you can create multiple counters named `methodCounter` that return the count of the method that is identified by the tag.
```
